source/common/http/utility.cc

```diff
@@ -756,6 +757,65 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
   encodeLocalReply(is_reset, std::move(prepared_local_reply));
 }
 
+bool Utility::remoteAddressIsTrustedProxy(
+    const Envoy::Network::Address::InstanceConstSharedPtr& remote,
+    const std::vector<Network::Address::CidrRange> trusted_cidrs) {
+  for (const auto& cidr : trusted_cidrs) {
+    if (cidr.isInRange(*remote.get())) {
+      return true;
+    }
+  }
+  return false;
+}
+
+Utility::GetLastAddressFromXffInfo Utility::getLastNonTrustedAddressFromXFF(
+    const Http::RequestHeaderMap& request_headers,
+    const std::vector<Network::Address::CidrRange> trusted_cidrs) {
+  const auto xff_header = request_headers.ForwardedFor();
+  if (xff_header == nullptr) {
+    return {nullptr, false};
+  }
+
+  absl::string_view xff_string(xff_header->value().getStringView());
+  static constexpr absl::string_view separator(",");
+
+  const auto xff_entries = StringUtil::splitToken(xff_string, separator, false, true);
+  Network::Address::InstanceConstSharedPtr last_valid_addr;
+
+  for (auto it = xff_entries.rbegin(); it != xff_entries.rend(); it++) {
+    auto addr = Network::Utility::parseInternetAddressNoThrow(std::string(*it));
+    if (addr == nullptr) {
+      continue;
+    }
+    last_valid_addr = addr;
+
+    bool remoteAddressIsTrustedProxy = false;
+    for (const auto& cidr : trusted_cidrs) {
+      if (cidr.isInRange(*addr.get())) {
+        remoteAddressIsTrustedProxy = true;
+        break;
+      }
+    }
+
+    if (remoteAddressIsTrustedProxy) {
+      continue;
+    }
+
+    // If we reach here we found a non-trusted address
+    if (Network::Utility::isInternalAddress(*addr.get()) ||
+        Network::Utility::isLoopbackAddress(*addr.get())) {
+      // If the next IP after skipping all trusted IPs is internal or loopback
+      // we cannot determine the true remote IP: Either a proxy didn't add its IP,
+      // or XFF is malformed.
```

Comment:
sorry probably naive here but if every hop is trusted do we not trust the whole set?
Agree if we have trusted, untrusted, trusted I wouldn't trust the final trusted.

Reply:
The aim is to find the first non-trusted public IP. The condition here says if we treated all previous IPs as proxies we trust, but we then hit a private IP, we can't determine the "real" client IP so we bail out.

Reply:
I think in that case the "real" IP would be the private IP, right? I think that's a valid case, where the end client that the traffic originated from was an internal IP.

Reply:
The "real real" IP may well be a private IP, yes, but a private IP isn't useful to an Internet-facing proxy and, AFAIK, no implementation will consider private IPs. The exception I've seen is Apache's mod_remoteip which has a separate CIDR list for internal proxies.

If we want to consider allowing private IPs as being a real client IP I would prefer to gate that behind another setting.

Reply:
how about we at least call that out in the API docs.  I've done my fair share of telneting into localhost for debug so at least we can hopefully avoid surprising anyone that way.

Reply:
I another round of consideration. First, I was wrong on

> AFAIK, no implementation will consider private IPs

I couldn't remember where I got that from, so I went and checked a few implementations and none of them have a check like this. I suspect I got myself muddled over the general view that private IPs aren't useful (for Internet-facing proxies).

Second, on balance having this check is far too restrictive and makes assumptions about the use cases. As Alyssa alluded to, if we trust a proxy, we should trust that it validated the header.

Therefore, I've removed this block and updated the API docs to remove mention of that restriction.